File: source/materials/lambertian.cpp
#include "materials/lambertian.hpp"
#include "math/details.hpp"

lambertian::lambertian(colour const &colour_in) noexcept : super(colour_in) {}

auto lambertian::scatter(const ray &ray_in, const hit &hit_in) const -> std::optional<scatter_data>
{
   vec reflected = hit_in.normal + random_unit_vector();

   return std::make_pair(super::emission, ray(hit_in.position, reflected));
}

File: go-rt/materials/lambertian.go
package materials

import (
	"go_rt/core"
	"go_rt/maths"
)

type Lambertian struct {
    Diffuse core.Colour 
}

func (this Lambertian) Scatter(ray *core.Ray, hit *core.RayHit) ScatterData {
    random := maths.RandomNormalizedVec3()
    scatterDir := maths.Add(&hit.Normal, &random)

    if scatterDir.IsNearZero() {
        scatterDir = hit.Normal
    }

    return ScatterData{
        Emission: core.Colour{}, 
        Diffuse: this.Diffuse, 
        Ray: core.Ray{Origin: hit.Position ,Direction: scatterDir}}
}
